### PR TITLE
Add uploading files

### PR DESCRIPTION
**Pull request by isabloCh**
Added uploading files and displaying book covers

### CODE REVIEW

File: app/admin/books.rb [hunk 1]
```diff
@@ -1,9 +1,23 @@
...
+  permit_params :title, :description, :recommended_church, :price, image_attributes: [:cover, :book_id]
+
+  form :html => { :enctype => "multipart/form-data" } do |f|
+    f.inputs "Book Details" do
+      f.input :title
+      f.input :description
+      f.input :price
+      f.input :recommended_church
+      f.inputs 'Image', for: ['image', f.object.image || f.object.build_image] do |book_form|
+        fi.input :cover, as: :file, hint: (book_form.object.cover ? 'No image' : image_tag(book_form.object.cover.url(:thumb)))
```
Comment (resivalex) [inline app/admin/books.rb:15]: `fi` -> `book_form`

File: app/admin/books.rb [hunk 2]
```diff
@@ -12,12 +26,19 @@
...
+    column :created_at    
+    column "Cover" do |book|
+      book.image ? image_tag(book.image.cover.url(:thumb)) : 'No image'
+    end
     actions
   end
 
+  
+
+
```
Comment (resivalex) [inline app/admin/books.rb:38]: слишком щедр на пустые строки

File: app/admin/images.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  # has_many :users, through: :ordered_books
+
+  permit_params :cover, :book_id
+
+  
+
+  form :html => { :enctype => "multipart/form-data" } do |f|
+    f.inputs do
+      f.input :book_id
+      f.input :cover, :as => :file
```
Comment (resivalex) [inline app/admin/images.rb:14]: `:as => :file` не модно

модно `as: :file`

File: app/models/book.rb [hunk 4]
```diff
@@ -1,5 +1,8 @@
 class Book < ApplicationRecord
 
   has_and_belongs_to_many :users, through: :ordered_books
+  has_one :image
```
Comment (resivalex) [inline app/models/book.rb:4]: belongs_to :image

File: app/models/image.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class Image < ApplicationRecord
+
+  has_attached_file :cover, styles: { medium: '300x300#', thumb: '100x100#' }
+
+  # TODO:  Validation is disabled until there is an agreed format for book covers
+  do_not_validate_attachment_file_type :cover
```
Comment (resivalex) [inline app/models/image.rb:6]: напиши так, может быть, пригодится. `validates :picture, attachment_content_type: { content_type: // }`

File: app/models/image.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+class Image < ApplicationRecord
+
+  has_attached_file :cover, styles: { medium: '300x300#', thumb: '100x100#' }
```
Comment (resivalex) [inline app/models/image.rb:3]: `has_attached_file :picture, ...`

File: db/migrate/20180425142018_create_images.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class CreateImages < ActiveRecord::Migration[5.1]
+  def change
+    create_table :images do |t|
+      t.attachment :cover
```
Comment (resivalex) [inline db/migrate/20180425142018_create_images.rb:4]: `t.attachment :picture`

File: db/migrate/20180425142018_create_images.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+      t.belongs_to :book, index: true
```
Comment (resivalex) [inline db/migrate/20180425142018_create_images.rb:5]: предлагаю не завязываться на то, что картинка принадлежит книге

File: app/admin/books.rb [hunk 9]
```diff
@@ -1,9 +1,23 @@
...
+  permit_params :title, :description, :recommended_church, :price, image_attributes: [:cover, :book_id]
+
+  form :html => { :enctype => "multipart/form-data" } do |f|
+    f.inputs "Book Details" do
+      f.input :title
+      f.input :description
+      f.input :price
+      f.input :recommended_church
+      f.inputs 'Image', for: ['image', f.object.image || f.object.build_image] do |book_form|
```
Comment (resivalex) [inline app/admin/books.rb:14]:
```
cover = book_form.object.cover
fi.input :cover, as: :file, hint: (cover ? 'No image' : image_tag(cover.url(:thumb)))
```

**Review by resivalex — CHANGES_REQUESTED** [review]

File: Gemfile [hunk 10]
```diff
@@ -6,11 +6,13 @@ git_source(:github) do |repo_name|
 end
 
 gem 'activeadmin'
+gem 'activeadmin_addons'
```
Comment (Loremaster) [inline Gemfile:9]: I would recommend locking version of the gems you use. At least you can use `~>` although I prefer specific version.

File: Gemfile [hunk 11]
```diff
@@ -6,11 +6,13 @@ git_source(:github) do |repo_name|
...
 gem 'devise', '~> 4.4.1'
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
 gem 'rails', '~> 5.1.4'
 gem 'rubocop', '~> 0.52.1', require: false
 # Use postgresql as the database for Active Record
+gem 'paperclip', '~> 6.0.0'
```
Comment (Loremaster) [inline Gemfile:15]: Paperclip is deprecated and will not be supported anymore: https://robots.thoughtbot.com/closing-the-trombone

It's better to switch to carrierwave

File: app/admin/books.rb [hunk 12]
```diff
@@ -1,9 +1,23 @@
...
   # See permitted parameters documentation:
   # https://github.com/activeadmin/activeadmin/blob/master/docs/2-resource-customization.md#setting-up-strong-parameters
-  #
-  # has_many :users, through: :ordered_books
-  permit_params :title, :description, :recommended_church, :price
+
+
+  permit_params :title, :description, :recommended_church, :price, image_attributes: [:cover, :book_id]
+
+  form :html => { :enctype => "multipart/form-data" } do |f|
```
Comment (Loremaster) [inline app/admin/books.rb:8]: Can you use new hash syntax please? `html: { enctype: "multipart/form-data" }`

**Review by Loremaster — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 11 above)
Comment (resivalex) [reply Gemfile:15]: @isabloCh, 
image

**Pull request closed without merging**